crates/vm/src/expr.rs

```diff
@@ -1445,6 +1460,127 @@ impl QueryExpr {
         }
     }
 
+    /// Try to turn an inner join followed by a projection into a semijoin.
```

Comment:
FWIW, this rewrite might not be exercised by the subscription path, since the original plans should all be index joins.

Reply:
Previously, this logic was part of `try_index_join`; it's now its own method. We still get here for SQL semijoins, since the SQL compiler emits a `JoinInner` followed by a `Project`. It's just that `try_index_join` used to recognize that whole pattern.